File: 2807_insert_greatest_common_divisors_in_linked_list/insertGreatestCommonDivisors.go
package insertgreatestcommondivisorsinlinkedlist

// Definition for singly-linked list.

type ListNode struct {
	Val  int
	Next *ListNode
}

func insertGreatestCommonDivisors(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	prev := head
	curr := head.Next

	for curr != nil {
		val := gcd(prev.Val, curr.Val)
		node := &ListNode{Val: val, Next: curr}
		prev.Next = node
		prev = curr
		curr = curr.Next
	}

	return head
}

func gcd(a int, b int) int {
	for b != 0 {
		a, b = b, a%b
	}

	return a
}

File: 2807_insert_greatest_common_divisors_in_linked_list/insertGreatestCommonDivisors.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = head
        curr = head.next

        while curr is not None:
            gcd = self.gcd(prev.val, curr.val)
            node = ListNode(gcd, curr)
            prev.next = node
            prev = curr
            curr = curr.next
        
        return head
        
    def gcd(self, a: int, b: int):
        while b != 0:
            a, b = b, a % b

        return a